device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -1704,12 +1700,15 @@ private void addToWaitingQueue(IotHubTransportPacket packet)
             if (packet != null && packet.getMessage() != null && packet.getMessage().getCorrelatingMessageCallback() != null)
             {
                 Message message = packet.getMessage();
-                String messageId = message.getCorrelationId();
-                if (!correlationCallbacks.containsKey(messageId))
+                String correlationId = message.getCorrelationId();
+                if (correlationId != null && !correlationCallbacks.containsKey(correlationId))
                 {
-                    correlationCallbacks.put(messageId, message.getCorrelatingMessageCallback());
-                    correlationCallbackContexts.put(messageId, message.getCorrelatingMessageCallbackContext());
-                    correlationCallbacks.get(messageId).onRequestQueued(message, packet, correlationCallbackContexts.get(messageId));
+                    correlationCallbacks.put(correlationId, message.getCorrelatingMessageCallback());
+                    if (message.getCorrelatingMessageCallbackContext() != null)
```

Comment:
Consider calling and caching the result, so you don't have to call it twice.

Reply:
I don't see the point of adding this check. message.getCorrelationId(); can't return null. 
```java
    /**
     * Getter for the correlationId property
     * @return The property value
     */
    public String getCorrelationId()    
    {
        // Codes_SRS_MESSAGE_34_045: [The function shall return the message's correlation ID.]
        if (correlationId == null)
        {
            return "";
        }

        return correlationId;
    }
```

Reply:
Can you link to that? I am looking in our current master and that's not there.

Reply:
Ah I see, yes, that method is changed on my end (as part of a larger refactor) to a Lombok Getter/Setter but it isn't guarded with @NonNull. Thanks for catching that as I wasn't in the master branch when I was looking at the code.

I've updated the branch I'm working on to keep this behavior.